Dynamic scheduling based on plugins

nateyazdani opened the pull request:
This change allows statscache plugins to specify the exact frequency at which they fire and then generates at load-time the producers needed, which requires some corresponding [changes to statscache_plugins](https://github.com/fedora-infra/statscache_plugins/pull/6). In aim of this goal, I made the following modifications:
- Available plugin classes are loaded at start-up and are initialized by the specific producer to which they are attached (in contrast to how they are currently loaded and initialized all at once)
  - Similar to moksha, generators/iterables are now valid plugin entrypoints. In the future, I would like to add a utility function/class to automate generating a plugin for several frequencies, like how the "volume, by [stuff]" plugins do
- The cache buckets are taken out of `StatsConsumer` and placed in their respective producer objects. This removes the potential issue of needing the producers to all modify the consumer object.
- The `statscache.schdule.Schedule` and `statscache.schdule.Frequency` classes are replaced with a `statscache.plugins.Frequency` class (subjectively relocated because it's part of the plugin API). The new class has a slightly different purpose than it's predecessors. It's an extension of `datetime.timedelta`, and it gets much of the prior functionality from that (e.g., adding/subtracting for `.next()` \ `.prev()`) as well as some additional helpful properties, in particular hashability. All it was missing from the previous implementations was a similarly nice `str()` implementation, which has been added.
  - For this to work, the `BasePlugin.handle()` needed to have the timestamp of the cache empty-ing. A parameter for such a value was already present in the abstract method signature, so I assume that this was already intended.
  - Note that I was not 100% clear on what was happening in the `statscache.schedule` module, so it's entirely possible that I've misinterpreted some of what it does. Please don't hesitate to call me out :)
- There is also some assorted minor improvements that I made (e.g., `idx` -> `ident`) that I felt improved readability, but these are subjective and reversible if preferable.

Finally, I wanted to mention a difference in plugin scheduling behavior. Previously, all plugins were getting fired from each producer; if you look at the initialization output, all 7 default plugins were loaded (and fired) by each producer, meaning all plugins were being run on all frequencies. I don't think that this was the intention, especially since the message volume plugins had specialized versions for particular frequencies. However, if we would like to preserve this sort of behavior, where some or all plugins are run on multiple frequencies, that is something I can easily add.

Commit:
Create missing bucket for one-day frequency

Commit:
Change positional arguments given by keyword

The functions ``get_model()`` and ``get_plugin()`` from ``statscache.utils``
were being called with their final positional arguments supplied as keyword
arguments. Although it works just fine, it is redundant (``config=config``) and
a bit confusing.

Commit:
Separate plugin loading and initialization

Separate plugin loading and initialization so that the list of available plugins
is fixed at start-up and easily accessible without necessarily causing them to
be initialized. This also allows the list of plugin classes to be accessible
before the Moksha Hub configuration is even available. In addition, plugin
entry-points are now required to either be a specific plugin class or collection
of them. There is an associated commit for statscache_plugins.

Commit:
Rename ``plugin.idx`` to ``plugin.ident``

The ``idx`` property of the ``statscache.plugins.BasePlugin`` is a bit of a
misnomer; it's not really a numerical index, as its name might suggest, but
really a unique string identifier. I've tried to rename it appropriately.

Commit:
Relocate cache buckets to respective producers

Move the caches from the StatsConsumer to the owning StatsProducer. This gives us
greater flexibility in creating new producers, because the central consumer does
not need to be specialized for each new producer.

Commit:
Provide timestamp to handle() method of plugins

The abstract method signature of the method handle() on
statscache.plugins.BasePlugin includes a timestamp, but this is not actually
given when the method is invoked. This change adds that, and a corresponding
change in statscache_plugins corrects the method signature in the plugin
implementations.

Commit:
Simplify time lapse/frequency classes

Removes statscache.utils.{Schedule,Frequency} in favor of
statcache.plugins.Frequency, an extension of datetime.timedelta. This class
contain only the minimal amount of code necessary for our uses.

Commit:
Dynamically generate producers based on plugins

Replaces the static list of available producers with a class factory generator
that determines the frequencies desired by the available plugin classes and then
creates a producer class for each frequency. As part of this, producers are now
responsible for initializing their respective plugins and registering them in the
plugin table for access by the web app. This will facilitate merging the
initialize() method of plugins with the standard __init__(), since a database
session will now be available at the time of plugin instantiation.

ralphbean commented on statscache/producer.py:
```diff
@@ -13,81 +14,94 @@
-        self.sister = statscache.utils.find_stats_consumer(self.hub)
+        # set up the cache and connect to the consumer
+        self.cache = []
+        statscache.utils.find_stats_consumer(self.hub).register(self)
```
Nice, I like this approach.

Commit:
Update statscache.plugins.Frequency

Changes statscache.plugins.Frequency so that all frequencies are synchronized
at UTC midnight and introduces some associated methods.

nateyazdani commented:
The latest [commit](https://github.com/yazman/statscache/commit/dbe4835f3e27b5e0ca4df4c4d120f6ec506863bd) enhances the statscache.plugins.Frequency class such that it now synchronizes on UTC midnight, while preserving hashability.

ralphbean commented:
> The latest commit enhances the statscache.plugins.Frequency class such that it now synchronizes on UTC midnight, while preserving hashability.

It looks good to me after a first read through.  I wish I'd seen it earlier, but there's an old test suite in the tests/ dir for the old schedule stuff.  Could you update it for the new Frequency class for peace of mind and future-proofery?

Commit:
Synchronize producer/plugin frequencies

The statscache.plugins.Frequency class is repurposed as
statscache.frequency.Frequency to serve as a sort of synchronized timedelta.
The producer factory now initializes all instances of the class so that they
are all synchronized on the same moment, UTC midnight of the current day.
To faciltate this, plugin classes now include a timedelta instance as an
'interval' class attribute and receive the associated Frequency instance in the
__init__() method, which attaches it as a 'frequency' instance attribute.

Commit:
Update scheduling/frequency tests

Update the unit tests for the old statscache.schedule.{Frequency,Schedule} to be
used with the new statscache.frequency.Frequency.

nateyazdani commented:
@ralphbean done! Note that some of the test cases were specific to the old classes (that allowed irregular frequencies) and were removed entirely, and there are also some that I've added to try to cover the workings of the new class more thoroughly.

As discussed on IRC, synchronizing the longer frequencies to some extent (e.g., ensuring that a daily frequency always fires at 00:00 UTC) is a desirable behavior and has also been incorporated. As part of that, the way the plugin API interacts with this `Frequency` class is slightly changed. The end result is that plugin classes now include an `interval` attribute, which is a `datetime.timedelta` used by `statscache` proper to instantiate a `Frequency` class, which is passed along to the plugins.

ralphbean commented:
OK, that one default argument needs to be fixed and then this can be merged.  No need to wait for subsequent approval, you can just do it.

Commit:
Correct incorrect keyword parameter default

Moves time-dependent calls into the method body so that the values change with
time, as they are intended to.

Commit:
Simplify Frequency test case method calls

Replace redundant calls to Frequency methods with a single cast, where
appropriate. This could not be done before because freezegun could not affect
the default keyword-argument values as required for these tests to work.

nateyazdani commented:
Ah, I knew something felt wrong when I wrote that, but I just couldn't put my finger on it. This also explains why some calls from the unit tests didn't seem to be affected by `freezegun`. Thanks!
